File: xiaqianZhang/assignments/fastSlow/lc876/lc876.py
# Problem Statement #

# Given the head of a Singly LinkedList, write a method to return the middle node of the LinkedList.

# If the total number of nodes in the LinkedList is even, return the second middle node.

# Input: given the head of LL that doesn't contains cycle
# return the middle node of LL
# As we know about the fast and slow pointer, fast pointer is twice as fast as the slow pointer, 
#   when fast pointer reach the null, then the slow pointer is position at the middle node
# intialize fast and slow as the head 
# while fast and fast.next != null, we want to increment the fast and slow pointer
# until outside of the while loop, we return the slow pointer
# ex: 1-2-3-4-5-6 when the total number of nodes in this LL is even
#   fast: 1-3-5-null
#   slow: 1-2-3-4
# when fast is position at null, slow is position at 4, which is the second middle node

class Node:
  def __init__(self, value, next=None):
    self.value = value
    self.next = next


def find_middle_of_linked_list(head):
  slow = head
  fast = head
  while(fast != None and fast.next != None):
    fast = fast.next.next
    slow = slow.next
  return slow


def main():
  head = Node(1)
  head.next = Node(2)
  head.next.next = Node(3)
  head.next.next.next = Node(4)
  head.next.next.next.next = Node(5)

  print("Middle Node: " + str(find_middle_of_linked_list(head).value))

  head.next.next.next.next.next = Node(6)
  print("Middle Node: " + str(find_middle_of_linked_list(head).value))

  head.next.next.next.next.next.next = Node(7)
  print("Middle Node: " + str(find_middle_of_linked_list(head).value))


main()


# Solution
# -----
# def find_middle_of_linked_list(head):
#   slow = head
#   fast = head
#   while (fast is not None and fast.next is not None):
#     slow = slow.next
#     fast = fast.next.next
#   return slow



# -----
# Time complexity #
# The above algorithm will have a time complexity of O(N) where ‘N’ is the number of nodes in the LinkedList.

# Space complexity #
# The algorithm runs in constant space O(1).

File: serdaSimus/assignments/03FastSlowPattern/lc876/lc876.js
// Problem Statement #

// Given the head of a Singly LinkedList, write a method to return the middle node of the LinkedList.

// If the total number of nodes in the LinkedList is even, return the second middle node.

/*
  input: head pointer to a linked-list
  output: pointer that's at the middle node.

  Sol:
  Using a two pointer system, 1 slow, 1 fast, we can find the middle item, 
  by moving the two pointers slow at half the speed of fast.
  therefore, if fast is at 10, slow is at 5, or if fast is at 101, slow will be at 51.
  so by return slow when fast reaches the end of the linked list
  we would have solved the problem.

  while (fast is not null)
    move fast by 2
    move slow by 1
    check if (fast.next is null) (meaning we have reached the end of the l.list)
      if true, return slow.
  return head or null

  edge cases: head points to null, we should return head.
  edge cases: head.next points to null, 1 node, we return the node.

*/
class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

const find_middle_of_linked_list = function (head) {
  // TODO: Write your code here
  if (head === null) return null;
  if (head.next === null) return null;

  let moveBy1Ptr = head,
    moveBy2Ptr = head;

  // I am not clear on why the while loop has to be AND instead of OR
  while (moveBy2Ptr !== null && moveBy2Ptr.next !== null) {
    moveBy1Ptr = moveBy1Ptr.next;
    moveBy2Ptr = moveBy2Ptr.next.next;
  }

  return moveBy1Ptr;
};
// let head = new Node();
// console.log(`Middle Node: ${find_middle_of_linked_list(head)}`);

head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(5);

console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);

head.next.next.next.next.next = new Node(6);
console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);

head.next.next.next.next.next.next = new Node(7);
console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);

// Solution
// -----
// function find_middle_of_linked_list(head) {
//   let slow = head,
//     fast = head;
//   while ((fast !== null && fast.next !== null)) {
//     slow = slow.next;
//     fast = fast.next.next;
//   }
//   return slow;
// }

// -----

// Time complexity #
// The above algorithm will have a time complexity of O(N) where ‘N’ is the number of nodes in the LinkedList.

// Space complexity #
// The algorithm runs in constant space O(1).
